### ISSUE TITLE
One-time-link for passkey verification in cli

### ISSUE THREAD

**Issue by leoleoasd** [kanidm/kanidm#1703]
**Is your feature request related to a problem? Please describe.**

If I ssh into a remote machine, I can't use passkey authentication since I can't put my yubikey there.

**Describe the solution you'd like**

The cli would be able to talk to the server and generate a one-time link. User open that  link in their local machine's browser and finish the webauthn process. Cli polls the server for status update.

**Comment by Firstyear** [comment]
we are currently going to pursue this with the ssh-sk key types rather than this method. We have however talked about this method as oauth2 device authorisation as an alternative too.

It's also worth pointing out that "passkeys" are not a clear term - there are multiple definitions of what a passkey is, and some of those definitions are unusable with ssh for example. Most likely here you are referring to fido/ctap keys I assume.

**Comment by leoleoasd** [comment]
Yes I mean webauthn.
You mean kanidm server would allow authenticate via a ssh key? So I can forward my local ssh agent to the server and use kanidm client there?

**Comment by Firstyear** [comment]
I'm confused by what you mean here. Let's be specific. 

When you ssh to another machine, Kanidm is involved in distributing ssh keys to the machine, and providing rules via pam to allow/deny the user, and resolve the user via nsswitch. 

However, the Kanidm *client* tool running on that machine won't be able to use your ssh keys for client auth. It *may* be possible in the future, but I think it would be reasonably complex and I'm not sure of the value.

Is there a reason you can't use the kanidm client tool from your local machine ?

**Comment by leoleoasd** [comment]
I'm not talking about the authentication during the ssh key distributing or ssh process. I'm talking about running `kanidm login` on remote machines.

My local machine is windows and I have to use WSL. I don't have any environment installed on windows.
I think there are some other use cases, for example, if the kanidm server is in a private network and I have to ssh into a bastion machine to use kanidm, and I can't put my passkey there.

**Comment by yaleman** [comment]
Basically, #1523

**Comment by Firstyear** [comment]
Yes but then wed need the cli to accept oauth2 device code, which also seems unfun given that the cli elsewhere has to provide device code. So wed have a weird "do you want hevice code" when you don't and "do you want webauthn" when you don't. We don't get a clear way to auth without attempting weird HW detection.

**Comment by Firstyear** [comment]
Actually thinking further, it's worse - we actually can't do this with device code for re-authentication because it's not actually proving access to the original and bound credentials of the session again. So device code would have to be limited to RO actions, which doesn't really help much. So we'd need a completely different auth mechanism that allows ssh agent forwarding/keys, or we need to come up with something seperate and unrelated. But device code won't cut it here IMO.

**Comment by dvv** [comment]
Hmm. So there is no way yet to login as a person with the only google authenticator passkey at a remote host, right?

**Comment by yaleman** [comment]
For SSH? no

**Comment by dvv** [comment]
Am sorry @yaleman, but all these "sshs" and "logins" might shadow the key sense.
To be precise, if a person XXX has registered only google authenticator passkey, that person can not succeed in running `kanidm person get XXX -D XXX`, right?

**Comment by yaleman** [comment]
Correct, only hardware tokens are currently supported on the CLI.

**Comment by dvv** [comment]
Same for password+TOTP on the CLI?

**Comment by Firstyear** [comment]
Yes, because they use a different mechanism. The current linux model assumes you use ssh key auth only for initial login, the pw to sudo.

**Comment by dvv** [comment]
Hmm.
```
$ kanidm login -D z11
Enter TOTP: xxxyyy
Enter password:
Login Success for z11@example.com
```
That's good.

**Comment by Firstyear** [comment]
Yes, but that's using a different mechanism than authentication over ssh. The context here is that the authentication protocols available over SSH and on linux are limited, because linux authentication is stuck in the stone ages.

**Comment by dvv** [comment]
I see. It's not a problem -- I'm just exploring the boundaries of _the given_ world.

**Comment by yaleman** [comment]
Ok, this thread got a little confused.

Once you've SSH'd to a remote host, TOTP authentication to the `kanidm` tool *will* work because it's just text entry. 

Anything more complex (ie, passkeys, security keys, CTAP, etc) won't because they require communications (USB, Bluetooth etc) that we can't make work through an SSH terminal session.

**Comment by dvv** [comment]
What about google authenticator? Once I've SSH'd to a remote host, can it be a stdout-printed link that is opened in the local browser and that eventually hits the authentication started in a remote ssh session?

**Comment by yaleman** [comment]
Not currently, and see above for more discussion on that point.
